### PR TITLE
feat(spool): Introduce in-memory `EnvelopeBuffer` as alternative to the old spooler

### PR DESCRIPTION
**Pull request by jjbayer**
This PR introduces a new envelope buffer as designed in https://github.com/getsentry/team-ingest/issues/345, though which all envelopes flow before being processed. Buffered envelopes are stored in per-project stacks. Stacks are sorted by whether the project configs needed for processing are ready, and by latest receive time, such that the latest envelope with a valid project config is always the first in line to be processed.

Currently, only memory-based envelope stacks are supported, but this PR already contains the implementation for disk-based stacks as well.

## Class Diagram (partial)

```mermaid
classDiagram




RequestHandler --> "writes to" GuardedEnvelopeBuffer
RequestHandler --> "requests update" ProjectCache


ProjectCache --> "notifies" GuardedEnvelopeBuffer

GuardedEnvelopeBuffer "1" *-- "1" PolymorphicEnvelopeBuffer

class PolymorphicEnvelopeBuffer {
    << enum >>
}

PolymorphicEnvelopeBuffer <|-- EnvelopeBuffer~StackProvider~

class EnvelopeBuffer {
    + push(envelope)
    + pop() Option~Envelope~
    + peek() Option~&Envelope~
    + mark_ready(ProjectKey)
    + mark_pending(ProjectKey)
}

EnvelopeBuffer "1" *-- "1" StackProvider
EnvelopeBuffer "1" *-- "1" PriorityQueue
PriorityQueue "1" *-- "n" EnvelopeStack

class StackProvider {
    << interface >>
    + create_stack()
}

StackProvider <|-- MemoryStackProvider
StackProvider <|-- SqliteStackProvider

StackProvider --> "creates" EnvelopeStack

class EnvelopeStack {
    <<interface>>
    + push(envelope)
    + pop() optional Envelope
}

EnvelopeStack <|-- MemoryEnvelopeStack

EnvelopeStack <|-- SqliteEnvelopeStack
```

ref: https://github.com/getsentry/team-ingest/issues/345

### CODE REVIEW

File: relay-config/src/config.rs [hunk 1]
```diff
@@ -842,6 +842,16 @@ fn spool_envelopes_unspool_interval() -> u64 {
     100
 }
 
+/// Default batch size for the stack.
+fn spool_envelopes_stack_disk_batch_size() -> usize {
+    1000
```
Comment (jjbayer) [inline relay-config/src/config.rs]: @iambriccardo should we use the same number as the old spooler here?

https://github.com/getsentry/relay/blob/b6262f27dec65ca5df21f6d3b8764796819904a5/relay-server/src/services/spooler/mod.rs#L69-L70

File: relay-server/src/services/cogs.rs [hunk 2]
```diff
@@ -1,9 +1,10 @@
 use std::sync::atomic::{AtomicBool, Ordering};
 
+use sentry_usage_accountant::{Producer, UsageAccountant, UsageUnit};
```
Comment (jjbayer) [inline relay-server/src/services/cogs.rs]: nit: We treat external dependencies the same as other relay crates for import order, so this was in the right place IMO.

File: relay-server/src/services/metrics/aggregator.rs [hunk 3]
```diff
@@ -1,15 +1,16 @@
-use hashbrown::HashMap;
 use std::time::Duration;
 
+use hashbrown::HashMap;
+
```
Comment (jjbayer) [inline relay-server/src/services/metrics/aggregator.rs]:
```suggestion
```

File: relay-server/src/services/project/state/info.rs [hunk 4]
```diff
@@ -1,4 +1,8 @@
 use chrono::{DateTime, Utc};
+use serde::{Deserialize, Serialize};
+use smallvec::SmallVec;
+use url::Url;
```
Comment (jjbayer) [inline relay-server/src/services/project/state/info.rs]: @iambriccardo See comment above.

File: relay-config/src/config.rs (hunk 1 above)
Comment (iambriccardo) [reply relay-config/src/config.rs]: We could yes, if you want, feel free to update them!

File: relay-server/src/services/cogs.rs (hunk 2 above)
Comment (iambriccardo) [reply relay-server/src/services/cogs.rs]: It's the automatic imports formatting that messes with this. Will fix

File: relay-server/src/services/project_cache.rs [hunk 5]
```diff
@@ -1010,16 +1042,118 @@ impl ProjectCacheBroker {
...
+                ProcessingGroup::Ungrouped,
+            );
+            managed_envelope.reject(Outcome::Invalid(DiscardReason::Timestamp));
+            // TODO: metrics in all branches.
+            return Ok(());
+        }
+        let sampling_key = envelope.sampling_key();
+        let services = self.services.clone();
+
+        let project_key = envelope.meta().public_key();
```
Comment (iambriccardo) [inline relay-server/src/services/project_cache.rs]:
```suggestion
        let own_key = envelope.meta().public_key();
```

File: relay-server/src/services/project_cache.rs [hunk 6]
```diff
@@ -1010,16 +1042,118 @@ impl ProjectCacheBroker {
...
+                peek.mark_ready(&sampling_key, false);
+                return Ok(());
+            }
+            None => None,
+        };
+
+        let project = self.get_or_create_project(project_key);
+
+        // Reassign processing groups and proceed to processing.
+        for (group, envelope) in ProcessingGroup::split_envelope(*peek.remove().await?) {
```
Comment (iambriccardo) [inline relay-server/src/services/project_cache.rs]:
```suggestion
        // Reassign processing groups and proceed to processing.
        let envelope = *peek.remove().await?;
        for (group, envelope) in ProcessingGroup::split_envelope(envelope) {
```

**Review by iambriccardo — APPROVED** [review]
LGTM, reviewed offline. Will discuss next steps

File: relay-config/src/config.rs [hunk 7]
```diff
@@ -868,6 +882,40 @@ pub struct EnvelopeSpool {
...
+    #[serde(default = "spool_envelopes_stack_max_batches")]
+    max_batches: usize,
+    /// Maximum time between receiving the envelope and processing it.
+    ///
+    /// When envelopes spend too much time in the buffer (e.g. because their project cannot be loaded),
+    /// they are dropped. Defaults to 24h.
+    #[serde(default = "spool_envelopes_max_envelope_delay_secs")]
+    max_envelope_delay_secs: u64,
+    /// Version of the spooler.
+    #[serde(default = "EnvelopeSpoolVersion::default")]
```
Comment (loewenheim) [inline relay-config/src/config.rs]: Nit:
```suggestion
    #[serde(default)]
```

File: relay-server/src/services/buffer/mod.rs [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+    /// From the tokio docs:
+    ///
+    /// >  The primary use case for the async mutex is to provide shared mutable access to IO resources such as a database connection.
+    /// > [...] when you do want shared access to an IO resource, it is often better to spawn a task to manage the IO resource,
+    /// > and to use message passing to communicate with that task.
+    backend: tokio::sync::Mutex<PolymorphicEnvelopeBuffer>,
+    /// Used to notify callers of `peek()` of any changes in the buffer.
+    notify: tokio::sync::Notify,
+    /// Used to notify callers of `peek()` of any changes in the buffer.
+    changed: AtomicBool,
```
Comment (loewenheim) [inline relay-server/src/services/buffer/mod.rs]: These fields have the same documentation, is that intentional?

Comment (jjbayer) [reply relay-server/src/services/buffer/mod.rs]: Yes, both are used for notification. I didn't find a good way to make it work with only a `Notify`.

**Pull request merged**